lib/logstash/pipeline.rb

```diff
@@ -277,15 +259,12 @@ def shutdown
         # input teardown must be synchronized since is can be called concurrently by
         # the input worker thread and from the pipeline thread shutdown method.
         # this means that input teardown methods must support multiple calls.
-        @run_mutex.synchronize{input.teardown}
+        @run_mutex.synchronize{input.stop}
       rescue LogStash::ShutdownSignal
```

Comment:
after re-analyzing the `@run_mutex` related code it is useful for 2 things:
- consistent view of the `@ready` and `@started` properties which need to be consistent across threads since the related [`ready?` and `started?` methods](https://github.com/jsvd/logstash/blob/79d822a03929c5dd30a6b92bcf81a1a8705b18b3/lib/logstash/pipeline.rb#L54-L60) will be typically called from another thread that the one executing the `run` method 
- to avoid calling the shutdown code before the startup is completed

I believe the `ready?` and `started?` methods could be implemented using atomic booleans and for the prevention startup and shutdown race condition it should be refactored in a better way. It really isn't clear now.

I will open another issue for this